UefiCpuPkg/CpuDxe/CpuMp.c

```diff
@@ -689,8 +694,24 @@ InitializeMpExceptionStackSwitchHandlers (
   }
 
   if (BufferSize != 0) {
-    Buffer = AllocateRuntimeZeroPool (BufferSize);
-    ASSERT (Buffer != NULL);
+    // we are allocating the buffer that will hold the new GDT and IDT for the APs. These must be allocated below
+    // 4GB as they are used by protected mode code on the APs when they are started up after this point. If they are
+    // above 4GB, the APs will triple fault because the 32 bit code segment is invalid
+    Buffer = (UINT8 *)(UINTN)(BASE_4GB - 1);
+    Status = gBS->AllocatePages (
+                    AllocateMaxAddress,
+                    EfiRuntimeServicesData,
+                    EFI_SIZE_TO_PAGES (BufferSize),
+                    (EFI_PHYSICAL_ADDRESS *)&Buffer
+                    );
+    if (EFI_ERROR (Status)) {
+      DEBUG ((DEBUG_ERROR, "Failed to allocate buffer for InitializeExceptionStackSwitchHandlers Status %r\n", Status));
+      ASSERT (FALSE);

```

Comment:
Current

```text
DXE_ASSERT!: [CpuDxe] c:\src\edk2\UefiCpuPkg\CpuDxe\CpuMp.c (709): ((BOOLEAN)(0==1))
```

Proposed

```text
ASSERT_EFI_ERROR (Status = Out of Resources)

DXE_ASSERT!: [CpuDxe] c:\src\edk2\UefiCpuPkg\CpuDxe\CpuMp.c (709): !(((RETURN_STATUS)(Status)) >= 0x8000000000000000ULL)
```

```suggestion
      ASSERT_EFI_ERROR (Status);
```

Reply:
Updated.